src/tokenizer.ts

```diff
@@ -1,105 +0,0 @@
-    },
-    one_line_comment: /\/\/.*?$/,
-    multi_line_comment_start: { match: /\/\*/, next: 'comment' },
-    number: /[0-9]+(?:\.[0-9]*)?(?![a-zA-Z_$[])/,
-    string_start_dquote: { match: '"', next: 'string_dquote' },
-    string_start_squote: { match: '\'', next: 'string_squote' },
-    boolean: /(?:[tT][rR][uU][eE]|[fF][aA][lL][sS][eE])(?![a-zA-Z_0-9])/,
-    identifier_prefix: /[[$](?!\s)/,
-    identifier_part: /[a-zA-Z_]\w*/,
-    identifier_suffix: /(?<!\s)]/,
-    lparen: '(',
-    rparen: ')',
-    comma: ',',
-    dot: '.',
-    bop: ['+', '-', '/', '*', '=', '==', '&&', '||', '&', '|'],
-    uop: '!',
-  },
-  string_dquote: {
-    escaped_char: /\\./,
-    string_contents: { match: /[^\\"]/, lineBreaks: true },
-    string_end_dquote: { match: '"', next: 'main' },
-  },
-  string_squote: {
-    escaped_char: /\\./,
-    string_contents: { match: /[^\\']/, lineBreaks: true },
-    string_end_squote: { match: '\'', next: 'main' },
-  },
-  comment: {
-    comment_end: { match: /\*\//, next: 'main' },
-    comment_contents: { match: /[^*]+/, lineBreaks: true },
-    comment_contents_asterisk: /\*(?!\/)/,
-  },
-})
-
-export const extractFormulaIdentifiers = (formula: string): string[] => {
-  const isIdentifierStart = (currentToken: { type?: string }): boolean => (
-    !!currentToken.type && ['identifier_prefix', 'identifier_part'].includes(currentToken.type)
-  )
-  const isIdentifierMiddleOrEnd = (currentToken: { type?: string }): boolean => (
-    !!currentToken.type && ['identifier_part', 'dot', 'identifier_suffix'].includes(currentToken.type)
-  )
-  const isIdentifier = (token: {type?: string}): boolean => (
-    !!token.type && ['identifier_prefix', 'identifier_part', 'dot', 'identifier_suffix'].includes(token.type)
-  )
-  const identifiers: string[] = []
-  formulaLexer.reset(formula)
-  let currentIdentifier = ''
-  /*
-  We need to distinguish between <identifier><whitespace><lparen> and <identifier><whitespace><identifier> - the first
-  is a function call and should be ignored whereas the second is two identifiers.
-  Since we read one token at a time, we don't know, when we read the whitespace, whether we should add an identifier to
-  the return value or not - we only know an identifier was finished.
-  */
-  let identifierEnded = false
-  for (const token of formulaLexer) {
-    if (identifierEnded && isIdentifier(token)) {
-      identifiers.push(currentIdentifier)
-      currentIdentifier = ''
-    }
-    identifierEnded = false
-
-    if (isIdentifierStart(token)) {
-      currentIdentifier += token.value
-    } else if (currentIdentifier && isIdentifierMiddleOrEnd(token)) {
-      currentIdentifier += token.value
-    }
-    if (currentIdentifier && token.type === 'whitespace') {
-      identifierEnded = true
-    } else if (currentIdentifier && !isIdentifier(token)) {
-      if (token.type !== 'lparen') {
-        // it's not a function call
-        identifiers.push(currentIdentifier)
```

Comment:
Correct me if I'm wrong, but you only allow one whitespace before an lparam to consider it a function.
What about something like `ISBLANK        (identifier)`? Not sure if this is allowed in Salesforce, but if one space is allowed there's a chance that more are allowed as well

Reply:
Ah, noticed that `whitespace` allows for at least one. I think this should work